clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -617,7 +618,7 @@ public boolean commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets,
 
             now = time.milliseconds();
             remainingMs = timeoutMs - (now - startMs);
-        } while (remainingMs > 0);
+        } while (remainingMs > 0 && !isFutureCompleted);
```

Comment:
Another way to see this bug is a failure to ensure completion of the future. Had we done so, then we could have skipped the `failed` check. This is why it worked prior to the patch which added the timeout. The pattern should really be something like this:
```java
if (future.isDone()) {
  if (future.succeeded()) {
    // handle success
  } else {
    // handle failure
  }
}
```
I guess one benefit of the enum approach is that it forces you to ensure completion prior to checking any of the possible results. That said, I'm a bit more inclined to remove the `isRetriable` method and leave it to the caller to determine what is and is not retriable. Then the request future only has two completion states.